rfc/20241021-dynamic-provider-instances.md

```diff
@@ -0,0 +1,766 @@
+					Key:      addrs.NoKey,
+				},
+				config,
+				EvalDataForNoInstanceKey,
+			)
+		case config.Alias != "" && config.ForEach != nil:
+			// Zero or more "additional" (aka "aliased") instances, with
+			// dynamically-selected instance keys.
+			instMap, moreDiags := evaluateForEachExpression(config.ForEach, modCtx)
+			diags = diags.Append(moreDiags)
+			if moreDiags.HasErrors() {
+				continue
+			}
+			for k := range instMap {
+				instKey := addrs.StringKey(k)
+				recordInst(
+					addrs.AbsProviderInstance{
+						Module:   moduleInstAddr,
+						Provider: n.addr.Provider,
+						Alias:    n.addr.Alias,
+						Key:      instKey,
+					},
+					config,
+					EvalDataForInstanceKey(instKey, instMap),
+				)
+			}
+		default:
+			// No other situation should be possible if the config
+			// decoder is correctly implemented, so this is just a
+			// low-quality error for robustness.
+			diags = diags.Append(fmt.Errorf("provider block has invalid combination of alias and for_each arguments; it's a bug that this wasn't caught during configuration loading, so please report it"))
+		}
+	}
+
+	addRootNodeToGraph(g)
+	return g, diags.ErrWithWarnings()
+}
+```
+
+The above preserves the existing concept of initially constructing an "abstract" node type and then passing it to a "concrete" function to wrap it in a specific implementation. During all phases except the special "eval" phase used for the `tofu console` command, the final node type is [`NodeApplyableProvider](https://pkg.go.dev/github.com/opentofu/opentofu@v1.8.3/internal/tofu#NodeApplyableProvider), whose `Execute` implementation is responsible for actually starting up the provider plugin (if it isn't a builtin) and calling `ConfigureProvider` on it. The resulting `providers.Interface` value is then stored in a map keyed by `addrs.AbsProviderInstance` so that other graph nodes can find it later.
+
+The other significant change during evaluation is in the implementation of the graph nodes representing individual resource instances. Because today's OpenTofu forces a one-to-one relationship between provider configuration and provider instance, the resource instance nodes already "know" exactly which provider instance they are bound to once the initial graph has been constructed. The implementation of these node types must now perform the final step of evaluating the optional "instance key" expression in the `provider` argument, to complete the information required to construct an `addrs.AbsProviderInstance` which can be used to look up the correct `providers.Interface` to use when performing the real operations related to that resource instance.
+
+#### State and Plan Format Changes
+
+The original state snapshot format version 4 tracks a provider configuration address for each entire resource, and then all of the instances of the resource are assumed to share that same provider configuration address. To support the features in this proposal requires two changes:
```

Comment:
This differs from #1911 in that we decided to store either the provider on the resource or on the instance.

If we decided to instead always write the resource.provider field, we should then only store the "provider instance key" in the resource instance.  That would make it easier for the code to be "correct" without additional validation.

The way we solved the problem of orphaned instances in #1911 was to include state provider references in NodeProviderAbstract.ProvidedBy.  This means that even though the config may have been removed, the data from the state will re-establish those links if necessary. I will admit its a bit of a hack as we had to defer some of the error handing to NodeResourcePlan.DynamicExpand, which it sounds like this proposal would be doing anyways.

Reply:
Assuming we do as I proposed here and also allow provider blocks inside module calls that use `count` and `for_each`, we also need to store the module instance address unless we invent some new serialization of that information.

(It is probably technically possible to infer it from the module instance path that the resource instance belongs to, but that seems quite fiddly and complicated compared to just using the address syntax we already use elsewhere and checking for consistency during load. There are already plenty other consistency checks during state snapshot loading.)